app/src/main/java/edu/nextstep/camp/calculator/MainPresenter.kt

```diff
@@ -2,39 +2,55 @@ package edu.nextstep.camp.calculator
 
 import edu.nextstep.camp.calculator.domain.Calculator
 import edu.nextstep.camp.calculator.domain.Expression
+import edu.nextstep.camp.calculator.domain.ExpressionHistorySaver
+import edu.nextstep.camp.calculator.domain.MemoryHistorySaver
 import edu.nextstep.camp.calculator.domain.Operator
 
 class MainPresenter(
     private val view: MainContract.View,
+    private val historySaver: ExpressionHistorySaver = MemoryHistorySaver(),
 ) : MainContract.Presenter {
 
     private val calculator = Calculator()
     private var expression = Expression.EMPTY
+    private var isHistoryDisplayed = false
 
     override fun addToExpression(operand: Int) {
         expression += operand
-        view.showExpression(expression)
+        view.showExpression(expression.toString())
     }
 
     override fun addToExpression(operator: Operator) {
         expression += operator
-        view.showExpression(expression)
+        view.showExpression(expression.toString())
     }
 
-    override fun removeLast() {
+    override fun removeLastFromExpression() {
         expression = expression.removeLast()
-        view.showExpression(expression)
+        view.showExpression(expression.toString())
     }
 
-    override fun calculate() {
+    override fun calculateExpression() {
         val result = calculator.calculate(expression.toString())
 
         if (result == null) {
             view.showCalculationFailMessage()
             return
         }
 
+        historySaver.save(expression.toString(), result)
         expression = Expression.newInstance(result)
-        view.showExpression(expression)
+        view.showExpression(expression.toString())
+    }
+
+    override fun toggleExpressionHistory() {
+        isHistoryDisplayed = !isHistoryDisplayed
+        if (isHistoryDisplayed) {
+            historySaver.loadHistories()
+                .map { (rawExp, result) -> CalculationHistoryItem(rawExp, result) }
+                .also { view.openCalculationHistories(it) }
+        } else {
+            view.closeCalculationHistories()
+        }
     }
```

Comment:
개인적으로는 if-else 문을 좋아하진 않습니다. if-return 문이나 코틀린은 훌륭한 when 문이 있으니 내부 로직을 함수로 Extract 한후 코드가독성을 높여보면 어떨까요? 🙏

```suggestion
    override fun toggleExpressionHistory() {
        when (isHistoryDisplayed) {
            true -> changeExpressionMode()
            false -> changeExpressionHistoryMode()
        }
    }
```